### PR TITLE
Fixed portal artifacts targeting the Ai

### PR DESCRIPTION
**Pull request by Gamer3107**
## About the PR

fixed the portal artifact targeting the Ai and people in containers as suggested in issue 32644

## Why / Balance

https://github.com/space-wizards/space-station-14/issues/32644

## Technical details

- the targeting now checks for any Mob with a MindContainer per HashSet
- some of them get excluded because of being in a container
- it then adds the mobs to a list of valid targets, but only if they have an active mind, and are not excluded
- if there are any targets, select a random one like bevor 
- continues as bevor
- logs in the adminloger if there was no valid target

## Requirements

- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes

**Changelog**

:cl:
- fix: Portal artifacts no longer target the AI or people within containers

### CODE REVIEW

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs [hunk 1]
```diff
@@ -21,22 +36,52 @@ public override void Initialize()
...
+        var mindMobs = new HashSet<Entity<MindContainerComponent, TagComponent>>();
 
-        var minds = new List<EntityUid>();
-        var mindQuery = EntityQueryEnumerator<MindContainerComponent, TransformComponent>();
-        while (mindQuery.MoveNext(out var uid, out var mc, out var xform))
+        _lookup.GetEntitiesOnMap<MindContainerComponent, TagComponent>(map, mindMobs);
+        foreach (var comp in mindMobs)
         {
-            if (mc.HasMind && xform.MapID == map)
-                minds.Add(uid);
```
Comment (metalgearsloth) [inline Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs:31]: Just add a isentityincontainer check here and remove all the other code.

**Comment by Everturning** [comment]
i dislike this because i think it's funny. also you can see it as the artifact messing with the cameras to change where the ai is looking. the crate thing is okay too ig

**Comment by ArtisticRoomba** [comment]
This bug just round removes the AI so I'd rather this bug be fixed than leave it in for 1 second of comedic effect.

**Comment by thebadman4662** [comment]
> This bug just round removes the AI so I'd rather this bug be fixed than leave it in for 1 second of comedic effect.

But crew and antags badly need AI counters right now. :trollface:

**Comment by Everturning** [comment]
wait it round removes AI okay thats not good nvm

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs (hunk 1 above)
Comment (Gamer3107) [reply Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs]: thanks for the hint. This new version is SO much better now.

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs [hunk 2]
```diff
@@ -21,22 +27,36 @@ public override void Initialize()
...
         var map = Transform(artifact).MapID;
-        var firstPortal = Spawn(artifact.Comp.PortalProto, _transform.GetMapCoordinates(artifact));
+        var validMinds = new List<EntityUid>();
+        var mindCont = new HashSet<Entity<MindContainerComponent>>();
 
-        var minds = new List<EntityUid>();
-        var mindQuery = EntityQueryEnumerator<MindContainerComponent, TransformComponent>();
-        while (mindQuery.MoveNext(out var uid, out var mc, out var xform))
+        _lookup.GetEntitiesOnMap<MindContainerComponent>(map, mindCont);
+        foreach (var mind in mindCont)
```
Comment (slarticodefast) [inline Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs:34]: Why did you change this? Just keep the `MapID==map` check below.

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs [hunk 3]
```diff
@@ -21,22 +27,36 @@ public override void Initialize()
...
-        _transform.SwapPositions(target, secondPortal);
+            //Manual position swapping, because the portal that opens doesn't trigger a collision, and doesn't teleport targets the first time.
+            _transform.SwapPositions(target, secondPortal);
 
-        _link.TryLink(firstPortal, secondPortal, true);
+            _link.TryLink(firstPortal, secondPortal, true);
+        }
+        else
+        {
+            _adminLogger.Add(LogType.Teleport, $"Portal Artifact {ToPrettyString(artifact.Owner)} failed to find valid Teleport target");
```
Comment (slarticodefast) [inline Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs:59]: Admin logs are for information relevant to admins, not for debugging. Remove this.

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs [hunk 4]
```diff
@@ -21,22 +27,36 @@ public override void Initialize()
...
+            if (mind.Comp.HasMind && !_container.IsEntityOrParentInContainer(mind))
+            {
+                validMinds.Add(mind.Owner);
+            }
         }
 
-        var target = _random.Pick(minds);
-        var secondPortal = Spawn(artifact.Comp.PortalProto, _transform.GetMapCoordinates(target));
+        //this would only be 0 if there were a station full of AIs and no one else
+        if (validMinds.Count != 0)
```
Comment (slarticodefast) [inline Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs:44]: just make this a guard statement

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs [hunk 5]
```diff
@@ -1,7 +1,10 @@
 using Content.Server.Xenoarchaeology.XenoArtifacts.Effects.Components;
 using Content.Server.Xenoarchaeology.XenoArtifacts.Events;
+using Content.Shared.Administration.Logs;
+using Content.Shared.Database;
```
Comment (slarticodefast) [inline Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/PortalArtifactSystem.cs:4]:
```suggestion
```
remove unused imports

**Review by slarticodefast — CHANGES_REQUESTED** [review]
While you are already are working on it, could you also fix line 49 to be
`_transform.SwapPositions(target, artifact.Owner);`
This line is supposed to switch the target player with the artifact manually, but currently it only moving the player towards the artifact and the portal gets ignored because it cannot be moved.

**Review by slarticodefast — APPROVED** [review]
Tested in-game and works as intended. Thanks for the bugfix and your first contribution :)

**Comment by metalgearsloth** [comment]
Good stuff.
